sklearn/metrics/pairwise.py

```diff
@@ -839,6 +844,173 @@ def cosine_distances(X, Y=None):
+    that may contain mixed types of data, using an implementation
+    of Gower formula.
+
+    Parameters
+    ----------
+    X : {array-like, pandas.DataFrame} of shape (n_samples, n_features)
+
+    Y : {array-like, pandas.DataFrame} of shape (n_samples, n_features), \
+        default=None
+
+    categorical_features : array-like of str, array-like of int, \
+            array-like of bool, slice or callable, default=None
+        Indexes the data on its second axis. Integers are interpreted as
+        positional columns, while strings can reference DataFrame columns
+        by name.
+        A callable is passed the input data `X` and can return any of the
+        above. To select multiple columns by name or dtype, you can use
+        :obj:`~sklearn.compose.make_column_selector`.
+
+        By default all non-numeric columns are considered categorical.
+
+    scale : bool, default=True
+        Indicates if the numerical columns should be scaled to [0, 1].
+        If ``False``, the numerical columns are assumed to be already scaled.
+        The scaling factors, _i.e._ ``min_values`` and ``scale_factor``, are
+        taken from ``X``. If ``X`` and ``Y`` are both provided, ``min_values``
+        and ``scale_factor`` have to be provided as well.
+
+    min_values : ndarray of shape (n_features,), default=None
+        Per feature adjustment for minimum. Equivalent to
+        ``min_values - X.min(axis=0) * scale_factor``
+        If provided, ``scale_factor`` should be provided as well.
+        Only relevant if ``scale=True``.
+
+    scale_factor : ndarray of shape (n_features,), default=None
+        Per feature relative scaling of the data. Equivalent to
+        ``(max_values - min_values) / (X.max(axis=0) - X.min(axis=0))``
+        If provided, ``min_values`` should be provided as well.
+        Only relevant if ``scale=True``.
+
+    Returns
+    -------
+    distances : ndarray of shape (n_samples_X, n_samples_Y)
+
+    References
+    ----------
+    Gower, J.C., 1971, A General Coefficient of Similarity and Some of Its
+    Properties.
+
+    Notes
+    -----
+    Categorical ordinal attributes should be treated as numeric for the purpose
+    of Gower similarity.
+
+    Current implementation does not support sparse matrices.
+
+    This implementation modifies the Gower's original similarity measure in
+    the sense that this implementation returns 1-S.
+    """
+    def _nanmanhatan(x, y):
+        return np.nansum(np.abs(x - y))
+
+    def _non_nans(x, y):
+        return len(x) - np.sum(_object_dtype_isnan(x) | _object_dtype_isnan(y))
+
+    def _nanhamming(x, y):
+        return np.sum(x != y) - np.sum(
+            _object_dtype_isnan(x) | _object_dtype_isnan(y))
+
+    if issparse(X) or issparse(Y):
+        raise TypeError("Gower distance does not support sparse matrices")
+
+    if X is None or len(X) == 0:
+        raise ValueError("X can not be None or empty")
+
+    if scale:
```

Comment:
I wonder whether it's worth running `locals().update(_precompute_metric_params(X, Y, 'gower', **locals())` to reduce code duplication... and move to a more elegant solution eventually?